### PR TITLE
Realm Implemented

### PR DESCRIPTION
**Pull request by Matias-Villalobo**
In this commit, Realm Database was implemented. Right now you can only see the stored data, through the Debugger

image

https://user-images.githubusercontent.com/73836200/111634476-517f1980-87d5-11eb-8024-5cdabe299298.mp4

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -58,4 +70,6 @@ dependencies {
     implementation 'com.android.support:recyclerview-v7'
     // Glide
     implementation 'com.github.bumptech.glide:glide:4.11.0'
+
```
Comment (SartiMau) [inline app/build.gradle:73]: remove this extra line pls

File: app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt [hunk 2]
```diff
@@ -14,5 +13,5 @@ interface MarvelApi {
     fun getCharacter(): Call<MarvelBaseResponse<MarvelResponse<List<CharacterResponse>>>>
 
     @GET("/v1/public/characters/{characterId}")
-    fun getSingleCharacter(@Path("characterId") id: Int): Call<MarvelBaseResponse<MarvelResponse<List<CharacterDetailResponse>>>>
+    fun getSingleCharacter(@Path("characterId") id: Int): Call<MarvelBaseResponse<MarvelResponse<List<CharacterEntity>>>>
```
Comment (BQuinteros) [inline app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt]: when I read entity, I think you are getting that CharacterEntity from domain. This character is we receive from service, maybe the correct name could be CharacterResponse

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt [hunk 3]
```diff
@@ -26,14 +25,15 @@ object CharacterMapper {
             characterResponse.description,
             characterResponse.modified,
             transformThumbnail(characterResponse.thumbnail),
+            characterResponse.modified,
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt]: why do you need this? this line is repeated in the line 26

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import com.example.marvelapp.domain.entity.EventsEntity
+import com.example.marvelapp.domain.entity.SeriesEntity
+import com.example.marvelapp.domain.entity.StoriesEntity
+import com.example.marvelapp.domain.entity.ThumbnailEntity
+import com.example.marvelapp.domain.entity.UrlsEntity
+import io.realm.RealmList
+
+object CharacterRealmMapper {
+
+    private fun transformDBCharacter(characterEntity: com.example.marvelapp.domain.entity.CharacterEntity): CharacterRealmEntity =
```
Comment (BQuinteros) [inline app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: Here if you put alt + enter, AS said you to put that com.example.marvelapp..etc in imports

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt [hunk 5]
```diff
@@ -26,14 +25,15 @@ object CharacterMapper {
...
             transformEvents(characterResponse.events),
             transformComics(characterResponse.comics),
             transformStories(characterResponse.stories),
             transformSeries(characterResponse.series),
             transformListOfUrls(characterResponse.urls)
         )
 
-    fun transformSingleCharacter(character: CharacterDetailResponse): CharacterDetailEntity =
+    fun transformSingleCharacter(character: com.example.marvelapp.data.service.response.CharacterEntity): CharacterDetailEntity =
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt]: check if you can avoid define all the path for the CharacterEntity

File: app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package com.example.marvelapp.data.local.model
+
+import io.realm.RealmList
+import io.realm.RealmObject
+import io.realm.annotations.PrimaryKey
+
+open class CharacterRealmEntity(
+    @PrimaryKey
+    var id: Int = 0,
```
Comment (BQuinteros) [inline app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt]: Here you can create a default const val... because for example if in the future I say the default value of response is gonna be "Not Value" i need to change it in a lot of places instead of you create a const, I change the value in only one place

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt (hunk 4 above)
Comment (BQuinteros) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: Check on other places pls

File: app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt (hunk 2 above)
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt]: why did you change this? this is not okay. The service should response something like CharacterDetailResponse

File: app/src/main/java/com/example/marvelapp/data/service/response/CharacterEntity.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
+package com.example.marvelapp.data.service.response
+
+open class CharacterEntity(
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/data/service/response/CharacterEntity.kt]: where do you use this? if it is a service response should have the **SerializedName** if not should be in domain

File: app/src/main/java/com/example/marvelapp/presentation/mvp/contract/MarvelAppContract.kt [hunk 8]
```diff
@@ -7,6 +7,8 @@ interface MarvelAppContract {
 
     interface MarvelAppModel {
         fun getHeroesData(): Observable<List<CharacterEntity>>
+        fun addHeroesToDataBase(data: List<CharacterEntity>)
+
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/presentation/mvp/contract/MarvelAppContract.kt]: remove this extra line pls

File: app/src/main/java/com/example/marvelapp/presentation/mvp/model/MarvelAppModel.kt [hunk 9]
```diff
@@ -2,12 +2,21 @@ package com.example.marvelapp.presentation.mvp.model
...
+    private val saveCharactersDataBaseUseCase: SaveCharactersDataBaseUseCase
+) : MarvelAppContract.MarvelAppModel {
 
     override fun getHeroesData(): Observable<List<CharacterEntity>> =
         getCharacterUseCases.call()
+
+    override fun addHeroesToDataBase(data: List<CharacterEntity>) {
+        saveCharactersDataBaseUseCase.call(data)
+    }
+
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/presentation/mvp/model/MarvelAppModel.kt]: remove these extra lines

File: app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt (hunk 2 above)
Comment (BQuinteros) [reply app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt]: You had a CharacterResponse, if it is correct use here pls

**Review by SartiMau — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/marvelapp/presentation/mvp/model/MarvelAppModel.kt (hunk 9 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/presentation/mvp/model/MarvelAppModel.kt]: Done thx

File: app/src/main/java/com/example/marvelapp/presentation/mvp/contract/MarvelAppContract.kt (hunk 8 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/presentation/mvp/contract/MarvelAppContract.kt]: Done thx

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            transformDBThumbnail(characterEntity.thumbnail),
+            characterEntity.resourceUri,
+            transformDBComics(characterEntity.comics),
+            transformDBEvents(characterEntity.events),
+            transformDBStories(characterEntity.stories),
+            transformDBSeries(characterEntity.series),
+            transformDBListOfUrls(characterEntity.urls) as RealmList<UrlRealmEntity>
+        )
+
+    fun transformDBListOfUrls(urlsEntity: List<UrlsEntity>): List<UrlRealmEntity> {
```
Comment (lucianoeli) [inline app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: cant you return a RealmList here and avoid de casting on line 32?

File: app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt (hunk 2 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt]: Thx for the advice, now its fixed

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt (hunk 4 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: Thx, now its fixed.

image

File: app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt (hunk 6 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt]: Thx, now its not hardcoded anymore

image

File: app/build.gradle (hunk 1 above)
Comment (Matias-Villalobo) [reply app/build.gradle:73]: Done, thx

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt (hunk 3 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt]: Thx, it was a typo mistake, instead of modified, the correct property was resourceUri

image

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt (hunk 5 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterMapper.kt]: Thx, now its fixed, and clearer

File: app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt (hunk 2 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/service/api/MarvelApi.kt]: Thx, now its fixed, and i am reusing the code that i have in CharacterResponse

File: app/src/main/java/com/example/marvelapp/data/service/response/CharacterEntity.kt (hunk 7 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/service/response/CharacterEntity.kt]: Thx, you were right, i already have one in the domain folder, and i did not use this. Now its gone.

File: app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt (hunk 10 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/mapper/CharacterRealmMapper.kt]: Thx, for the advice, now its done

**Comment by Matias-Villalobo** [comment]
lastTest

File: app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    var id: Int = VALUE_ZERO,
+    var name: String = EMPTY_STRING,
+    var description: String = EMPTY_STRING,
+    var modified: String = EMPTY_STRING,
+    var thumbnail: ThumbnailRealmEntity? = ThumbnailRealmEntity(EMPTY_STRING, EMPTY_STRING),
+    var resourceUri: String = EMPTY_STRING,
+    var comics: ComicsRealmEntity? = ComicsRealmEntity(VALUE_ZERO, EMPTY_STRING, VALUE_ZERO),
+    var events: EventsRealmEntity? = EventsRealmEntity(VALUE_ZERO, EMPTY_STRING, VALUE_ZERO),
+    var stories: StoriesRealmEntity? = StoriesRealmEntity(VALUE_ZERO, EMPTY_STRING, VALUE_ZERO),
+    var series: SeriesRealmEntity? = SeriesRealmEntity(VALUE_ZERO, EMPTY_STRING, VALUE_ZERO),
```
Comment (LuchoElissondoGlobant) [inline app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt]: you dont need to initialize these entities with default parameters when you are using default parameters in the constructor

File: app/src/main/java/com/example/marvelapp/domain/usecase/SaveCharactersDataBaseUseCase.kt [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+    fun call(data: List<CharacterEntity>)
+}
+
+class SaveCharactersDataBaseUseCaseImpl(private val localMarvelDataBase: LocalMarvelDataBase) :
+    SaveCharactersDataBaseUseCase {
+
+    override fun call(data: List<CharacterEntity>) {
+        localMarvelDataBase.saveCharacters(data)
+    }
+
```
Comment (LuchoElissondoGlobant) [inline app/src/main/java/com/example/marvelapp/domain/usecase/SaveCharactersDataBaseUseCase.kt]: remove extra line please

File: app/src/main/java/com/example/marvelapp/data/local/model/utils/NumbersUtils.kt [hunk 13]
```diff
@@ -0,0 +1,5 @@
+package com.example.marvelapp.data.local.model.utils
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/data/local/model/utils/NumbersUtils.kt]: you can unify this class and the new StringUtils in a ContantsUtils

File: app/src/main/java/com/example/marvelapp/data/local/model/utils/StringUtils.kt [hunk 14]
```diff
@@ -0,0 +1,5 @@
+package com.example.marvelapp.data.local.model.utils
```
Comment (SartiMau) [inline app/src/main/java/com/example/marvelapp/data/local/model/utils/StringUtils.kt]: you can unify this class and the new NumbersUtils in a ContantsUtils

File: app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt (hunk 11 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/local/model/CharacterRealmEntity.kt]: Done, thx for the advice

File: app/src/main/java/com/example/marvelapp/domain/usecase/SaveCharactersDataBaseUseCase.kt (hunk 12 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/domain/usecase/SaveCharactersDataBaseUseCase.kt]: Done, thx

File: app/src/main/java/com/example/marvelapp/data/local/model/utils/StringUtils.kt (hunk 14 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/local/model/utils/StringUtils.kt]: Done, now they are in the seme file, thx!

File: app/src/main/java/com/example/marvelapp/data/local/model/utils/NumbersUtils.kt (hunk 13 above)
Comment (Matias-Villalobo) [reply app/src/main/java/com/example/marvelapp/data/local/model/utils/NumbersUtils.kt]: Done, now they are in the seme file, thx!

**Review by SartiMau — APPROVED** [review]

**Review by BQuinteros — APPROVED** [review]

**Review by mauroLonghi — APPROVED** [review]
